### ISSUE TITLE
Reconsider use of `lint.dummy-variable-rgx` in `RUF045`

### ISSUE THREAD

**Issue by ntBre** [astral-sh/ruff#27095]
While trying to stabilize [implicit-class-var-in-dataclass (RUF045)](https://docs.astral.sh/ruff/rules/implicit-class-var-in-dataclass/#implicit-class-var-in-dataclass-ruf045) for 0.16, we noticed that the applying [lint.dummy-variable-rgx](https://docs.astral.sh/ruff/settings/#lint_dummy-variable-rgx) to class variables in this context didn't seem quite right. This seems out of the scope of the setting, and these variables also seem like they should be in scope for the rule.

It sounds like this was added just before merging the rule to reduce the number of ecosystem hits: https://github.com/astral-sh/ruff/pull/14349#pullrequestreview-2619427807 but only from 16 hits to 8.

**Comment by wahajahmed010** [comment]
Quick clarification question before this gets too far down the design path, because I think the right behavior depends on which "class variable" we're talking about.

Concretely, which of these should `lint.dummy-variable-rgx` *not* apply to under RUF045?

```python
from dataclasses import dataclass
from typing import ClassVar

@dataclass
class A:
    _x: int = 0            # (1) underscore-prefixed, dataclass field
    x: ClassVar[int] = 0    # (2) explicit ClassVar
    y: int = 0             # (3) plain dataclass field, RUF045 will flag

@dataclass
class B:
    _: int                 # (4) bare underscore dataclass field
```

I'd argue:
- (1) and (4) should be filtered out by RUF045 if `dummy-variable-rgx` matches (and *should* be ignored even when it doesn't, because underscore-prefixed names are an explicit "I'm hiding this" signal in Python).
- (2) should never be flagged (it has an explicit `ClassVar` annotation — there's nothing implicit to warn about).
- (3) is the only case RUF045 should actually target.

If (1)/(4) are getting caught today because the rule treats them like ordinary annotations, that would also be a separate bug from the `dummy-variable-rgx` interaction the issue mentions. Either way, a test fixture covering these four cases would be a useful first step before deciding where the filter belongs. Happy to open a PR with that fixture if it would help.

**Comment by MichaReiser** [comment]
The reasoning makes sense. I think there's a question of what the argument is for excluding private members. If that's the intention, then we should only exclude those rather than depending on the dummy variable regex to accidentally catch them.

I think opening a PR makes sense. It also allows us to see the ecosystem impact of the change

**Comment by ntBre** [comment]
From my reading of the docs, it certainly seems in the spirit of the rule to apply to private members. Dataclasses require the argument whether it's private or not:

```pycon
>>> from dataclasses import dataclass
>>> @dataclass
... class C:
...     _x: int
...
>>> C()
Traceback (most recent call last):
  File "<python-input-3>", line 1, in <module>
    C()
    ~^^
TypeError: C.__init__() missing 1 required positional argument: '_x'
```

and thus I'd argue that the same confusion is present whether it's private or not.    

Just a note, the dummy regex is currently relevant for unannotated cases only, which is what the issue is about. RUF045 never flags annotated cases because those aren't implicit class vars, only unannotated cases. So the rule doesn't apply to any of the cases 1-4 above. The question is whether we should flag cases like _this_:

```py
@dataclass
class C:
    _x = 1
```
